File: core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceInfo.java
package com.rudderstack.android.sdk.core;

import android.content.ContentResolver;
import android.content.Context;
import android.os.Build;
import android.provider.Settings;
import android.util.Pair;

import com.google.gson.annotations.SerializedName;
import com.rudderstack.android.sdk.core.util.Utils;

import static com.rudderstack.android.sdk.core.util.Utils.isOnClassPath;

import androidx.annotation.Nullable;

class RudderDeviceInfo {
    @SerializedName("id")
    @Nullable
    private String deviceId;
    @SerializedName("manufacturer")
    private String manufacturer = Build.MANUFACTURER;
    @SerializedName("model")
    private String model = Build.MODEL;
    @SerializedName("name")
    private String name = Build.DEVICE;
    @SerializedName("type")
    private String type = "Android";
    @SerializedName("token")
    private String token;
    @SerializedName("adTrackingEnabled")
    private Boolean adTrackingEnabled;
    @SerializedName("advertisingId")
    private String advertisingId;

    RudderDeviceInfo(String advertisingId, String token, boolean collectDeviceId, RudderPreferenceManager preferenceManager) {

        if (collectDeviceId) {
            this.deviceId = Utils.getDeviceId(RudderClient.getApplication());
        }

        // update the advertisingId value in persistence, if user specifies one again
        // if the user didn't pass any advertisingId, then try reading it from preferences
        if (advertisingId != null && !advertisingId.isEmpty()) {
            preferenceManager.saveAdvertisingId(advertisingId);
            this.advertisingId = advertisingId;
        } else {
            this.advertisingId = preferenceManager.getAdvertisingId();
        }
        this.adTrackingEnabled = (this.advertisingId != null);

        if (token != null && !token.isEmpty()) {
            this.token = token;
        }
    }

    @Nullable
    String getDeviceId() {
        return deviceId;
    }

    void setToken(String token) {
        this.token = token;
    }

    void setAdTrackingEnabled(boolean enabled) {
        this.adTrackingEnabled = enabled;
    }

    void setAdvertisingId(String advertisingId) {
        this.advertisingId = advertisingId;
        this.adTrackingEnabled = (this.advertisingId != null);
        if (RudderClient.getApplication() != null) {
            RudderPreferenceManager preferenceManager = RudderPreferenceManager.getInstance(RudderClient.getApplication());
            preferenceManager.saveAdvertisingId(advertisingId);
        }
    }

    void setAutoCollectedAdvertisingId(String advertisingId) {
        this.advertisingId = advertisingId;
    }

    String getAdvertisingId() {
        return this.advertisingId;
    }

    boolean isAdTrackingEnabled() {
        return this.adTrackingEnabled;
    }

    void clearAdvertisingId() {
        this.advertisingId = null;
        this.adTrackingEnabled = (this.advertisingId != null);
        if (RudderClient.getApplication() != null) {
            RudderPreferenceManager preferenceManager = RudderPreferenceManager.getInstance(RudderClient.getApplication());
            preferenceManager.clearAdvertisingId();
        }
    }
}

File: models/src/main/java/com/rudderstack/models/android/RudderDeviceInfo.kt
package com.rudderstack.models.android

import com.fasterxml.jackson.annotation.JsonProperty
import com.google.gson.annotations.SerializedName
import com.squareup.moshi.Json

class RudderDeviceInfo(
    /*RudderDeviceInfo(String advertisingId, String token) {
        this.deviceId = Utils.getDeviceId(RudderClient.getApplication());
        if (advertisingId != null && !advertisingId.isEmpty()) {
            this.advertisingId = advertisingId;
            this.adTrackingEnabled = true;
        }
        if (token != null && !token.isEmpty()) {
            this.token = token;
        }
    }*/
    @SerializedName("id")
    @JsonProperty("id")
    @Json(name = "id")
    val deviceId: String? = null,

    @SerializedName("manufacturer")
    @JsonProperty("manufacturer")
    @Json(name = "manufacturer")
    private val manufacturer: String /*= Build.MANUFACTURER*/,

    @SerializedName("model")
    @JsonProperty("model")
    @Json(name = "model")
    private val model: String /*= Build.MODEL*/,

    @SerializedName("name")
    @JsonProperty("name")
    @Json(name = "name")
    private val name: String /*= Build.DEVICE*/,

    @SerializedName("type")
    @JsonProperty("type")
    @Json(name = "type")
    private val type: String = "Android",

    @SerializedName("token")
    @JsonProperty("token")
    @Json(name = "token")
    private var token: String? = null,

    @SerializedName("adTrackingEnabled")
    @JsonProperty("adTrackingEnabled")
    @Json(name = "adTrackingEnabled")
    var isAdTrackingEnabled: Boolean? = null,
) {

    @SerializedName("advertisingId")
    var advertisingId: String? = null
    fun setToken(token: String?) {
        this.token = token
    }
}
